### PR TITLE
fix(select): variant attr, tests

### LINKED ISSUE #2664: select: label accessibility
## Description of the issue*

- `pf-listbox` should have an accessible label
- `pf-select`'s toggle button text should not be the item selected but something more descriptive

* Note this breaks with upstream API but is more accessible.

### Impacted component(s)

- `pf-listbox`
- `pf-select`

### Steps to reproduce

1. Go to [Select documentation](https://patternflyelements.org/components/select/)
2. Inspect the first example.
3. Check the accesisbility tree.

### Expected behavior

Listbox and toggle button should have descriptive labels.

### PR DESCRIPTION
**Pull request by bennypowers**
Closes #2664

## What I did

1. replace boolean attrs with upstream's `variant` attr
2. remove `case-sensitive` and `match-anywhere` attrs from select and listbox, use a `customFilter` prop instead.
3. adjust tests

# Notes to reviewers

please carefully review the spec files

### CODE REVIEW

File: elements/pf-select/pf-select.ts [hunk 1]
```diff
@@ -197,63 +198,36 @@ export class PfSelect extends LitElement {
...
-    return this._listbox?.selected;
-  }
-
-  /**
-   * list of values as comma separated list
-   */
-  get selectedList() {
-    return this.#valueTextArray.map(txt => txt.replace(',', '\\,')).join(', ');
+  private get hasChips() {
+    return this.variant.startsWith('typeahead');
```
Comment (nikkimk) [inline elements/pf-select/pf-select.ts:209]: not all typeahead options have chips

**Review by nikkimk — CHANGES_REQUESTED** [review]

File: elements/pf-select/pf-select.ts [hunk 2]
```diff
@@ -197,63 +198,36 @@ export class PfSelect extends LitElement {
   /**
    * whether select has badge for number of selected items
    */
-  get hasBadge() {
-    return this.#isMulti && this.selectedItemsDisplay === 'badge';
+  private get hasBadge() {
+    return this.variant === 'checkbox';
```
[the remark this answers is not in the document]
Comment (bennypowers) [reply elements/pf-select/pf-select.ts]: bd5c3037

File: elements/pf-select/pf-select.ts (hunk 1 above)
Comment (bennypowers) [reply elements/pf-select/pf-select.ts]: bd5c3037

**Review by zeroedin — COMMENTED** [review]
https://deploy-preview-2677--patternfly-elements.netlify.app/components/select/demo/typeahead-chips/ Focus is doing something weird in this demo, clicking on input should allow you to type, instead focus moves to the listbox.

File: elements/pf-select/demo/typeahead-multiple.html [hunk 3]
```diff
@@ -1,5 +1,5 @@
 <label for="pfselect">Pick a color:</label>
-<pf-select id="pfselect" typeahead multi>
+<pf-select id="pfselect" variant="typeaheadmulti" multi>
```
Comment (bennypowers) [inline elements/pf-select/demo/typeahead-multiple.html:2]:
```suggestion
<pf-select id="pfselect" variant="typeaheadmulti">
```

**Review by zeroedin — COMMENTED** [review]
@nikkimk  I'm making assumptions that `demo/chips.html` and `demo/multi-select-with-badge.html` can now be removed?  Their attributes are no longer available.   Does that seem correct to you?

**Comment by nikkimk** [comment]
> @nikkimk I'm making assumptions that `demo/chips.html` and `demo/multi-select-with-badge.html` can now be removed? Their attributes are no longer available. Does that seem correct to you?

Docs and demos should align of upstream docs imho.

**Comment by zeroedin** [comment]
> Docs and demos should align of upstream docs imho.

My mistake I misunderstood what the demos were supposed to show.  Compared to PFv4 and have a better understanding now of what should be shown there.

**Comment by bennypowers** [comment]
> Options count for checkbox variant doesn't update on keyboard navigation selection, only on with mouse click.
> 
> image

2d39d804

**Comment by bennypowers** [comment]
@nikkimk If you write out the specs in prose, I can fill out the test file, e.g.

```
<pf-select> > variant attribute > "single" > calling focus()) > pressing Enter > opens
<pf-select> > variant attribute > "single" > calling focus()) > pressing Enter > pressing ArrowDown > focuses on first option
<pf-select> > variant attribute > "single" > calling focus()) > pressing Enter > pressing ArrowDown > pressing ArrowDown > focuses on second option
<pf-select> > variant attribute > "single" > calling focus()) > pressing Enter > pressing ArrowDown > pressing ArrowDown > pressing Enter > selects it
<pf-select> > variant attribute > "single" > calling focus()) > pressing Enter > pressing Space > closes
```
but for all the different scenarios

you're also welcome to take a crack at it, just commit whatever you have by eow so i can pick it back up sunday morning :handshake:

**Comment by nikkimk** [comment]
@bennypowers:

### Single 
- <pf-select> > variant attribute > "single" > calling focus()) > pressing Enter > opens and focuses on active item
- <pf-select> > variant attribute > "single" > calling focus()) > pressing Spacebar > opens and focuses on active item
- <pf-select> > variant attribute > "single" > calling focus()) > pressing ArrowDown > opens and focuses on active item
- <pf-select> > variant attribute > "single" > opened > first item focus() > pressing Shift+Tab > closes and sets focus on trigger
- <pf-select> > variant attribute > "single" > opened > first item focus() > pressing ArrowDown > focuses on second item
- <pf-select> > variant attribute > "single" > opened > second item focus() > pressing ArrowUp > focuses on first item
- <pf-select> > variant attribute > "single" > opened > first item focus() > pressing ArrowUp > focuses on last item
- <pf-select> > variant attribute > "single" > opened > last item focus() > pressing ArrowDown > focuses on first item
- <pf-select> > variant attribute > "single" > opened > first item focus() > pressing Enter > closes, selects activeItem, and sets focus on trigger

### Checkbox 
- <pf-select> > variant attribute > "checkbox" > calling focus()) > pressing Enter > opens and focuses on active item
- <pf-select> > variant attribute > "checkbox" > calling focus()) > pressing Spacebar > opens and focuses on active item
- <pf-select> > variant attribute > "checkbox" > calling focus()) > pressing ArrowDown > opens and focuses on active item
- <pf-select> > variant attribute > "checkbox" > opened > first item focus() > pressing Shift+Tab > closes and sets focus on 
- <pf-select> > variant attribute > "checkbox" > opened with second item selected > first item focus() > pressing Enter > adds activeItem to selection but does not close
- <pf-select> > variant attribute > "checkbox" > selecting two items > shows badge with number 2
- <pf-select> > variant attribute > "checkbox" > hide-badge attribute > true > selecting two items > no badge is shown

### Typeahead 
- <pf-select> > variant attribute > "typeahead" > calling focus()) > pressing ArrowDown > opens, focuses on active item, selects it, and updates input
- <pf-select> > variant attribute > "typeahead" > opened > second item focus() > pressing Enter >  selects it, updates input, focuses on input, and closes

### Typeahead multiple 
- <pf-select> > variant attribute > "typeaheadmulti" > calling focus()) > pressing ArrowDown > opens and focuses on active item
- <pf-select> > variant attribute > "typeaheadmulti" > opened > first item focus() > pressing Shift+Tab > closes and sets focus on 
- <pf-select> > variant attribute > "typeaheadmulti" > opened with second item selected > first item focus() > pressing Enter > adds activeItem to selection but does not close
- <pf-select> > variant attribute > "typeaheadmulti" > has-chips attribute??? > selecting first item > adds a chip and sets focus on input
- <pf-select> > variant attribute > "typeaheadmulti" > has-chips attribute??? > multiple items selected > selecting first chip > pressing Enter > deletes chip and focuses on next chip 
- <pf-select> > variant attribute > "typeaheadmulti" > has-chips attribute??? > one item selected > selecting first chip > pressing Enter > deletes chip and focuses on next input 
- <pf-select> > variant attribute > "typeaheadmulti" > has-chips attribute??? > no items selected > no chips displayed

**Comment by nikkimk** [comment]
@bennypowers per your question this morning for all select variants, when any listbox items is focused:

- pressing `Escape` closes the listbox and focuses on the toggle button
- pressing `Shift+Tab` closes the listbox and focuses on the toggle button
- pressing `Tab` closes the listbox and focuses on the next focusable element in the DOM

**Comment by bennypowers** [comment]
ok tests on pf-select should be green now. I think we might be missing some assertions on accessible labels and aria descriptions, it would be good to ensure those are there

As well, it seems to me that delegatesFocus has different behaviour in chrome vs firefox. focusing the dropdown and pressing arrowdown in chromium focuses the first option, but in firefox it does not.

**Review by nikkimk — CHANGES_REQUESTED** [review]
Docs page in DP is not correct.

**Comment by bennypowers** [comment]
@nikkimk not correct in what way?

**Comment by hellogreg** [comment]
Nikki may be referring to how the variant demos from "Multi-selectable" on down on the docs page are currently all just the default "Single" select. For example, this is how "Checkbox" looks:

Screencap of what should be the Checkbox select, which is really just a Single select

**Comment by hellogreg** [comment]
Also the "Accessibility" heading is currently a `<p>`: 
Accessibility section heading text

**Comment by hellogreg** [comment]
### Screen reader issues as of 1/31/24
It looks like Select is working as expected via keyboard in desktop browsers--when no screen reader is enabled. However, there are currently issues among all the major Mac and Win browser and screen reader combos (Mac Safari and VoiceOver, Win Chrome and JAWS, Win Firefox and NVDA).

The following can be tested/confirmed in the Single variant of Select (among others):

```
GIVEN a closed <pf-select> with keyboard focus

WHEN the user presses space or enter to open the list
  And then uses the arrow keys to highlight an option
  And then presses space or enter to select an option

THEN the highlighted option should be selected
  And the list should close 
```

### Mac Safari and VoiceOver

Right now, with Safari and VoiceOver, I _can_ open the list with space/enter. And I _can_ navigate the options with the up and down arrows.

However, pressing space/enter to select the highlighted option does not select the highlighted option. The first time you open the list and press space/enter, the list closes with the first option selected (e.g., "Blue" in the Single variant demo)--no matter which option you have highlighted. Upon subsequent openings of the list, the space/enter keys do nothing at all. The list can only be closed via tabbing or the escape key.

### Win Chrome/JAWS and FireFox NVDA

Though the list can be opened via space/enter, the lists cannot be navigated via the arrow keys at all.

_Hope that test scenario format is the kind of thing you're looking for, @bennypowers !_

**Comment by bennypowers** [comment]
I noticed some issues with the typeahead multi variant, as well. It could be that the firefox-focus-fix ... broke (😉) the behaviour when expanding the menu for typeaheadmulti

**Comment by bennypowers** [comment]
i've got a couple more changes i'm planning for this: ListboxController will have to receive a secondary keyboard a11y controller (either rtic or activedescendant) in order to function. this will mean small changes to the interface on rtic

**Comment by bennypowers** [comment]
@nikkimk, @hellogreg, and @zeroedin this is ready for a comprehensive review. Notes:
- Expect tabs tests to fail, we'll address those in a separate PR (per discussion today at office hours)
- Please validate the test cases as test files have changed
- Please consider downstream use through the lens of readability and maintainability. See this PR as an effort to limit the scope of the abstractions we offer in core. It might mean more work for downstream, but at the benefit of readability/maintainability/team-grokkability. Later on, when we have more downstream cases, we can always identity higher level abstractions, but if we make even small mistakes in high level abstractions now, we won't be able to support new features without breaking changes

**Review by nikkimk — CHANGES_REQUESTED** [review]
- Chip group should be hidden if there are not chips, but it can be seen in DP
-

File: elements/pf-select/pf-select.ts [hunk 4]
```diff
@@ -112,406 +109,336 @@ export class PfSelect extends LitElement {
...
+    //     this._chipGroup.focusOnChip(chip);
+    //     this._input.value = '';
+    //   }
+    // }
   }
 
   firstUpdated() {
-    this.#updateValueText();
+    // TODO: don't do filtering in the controller
+    // if (this.variant === 'typeaheadmulti') {
```
[the remark this answers is not in the document]
Comment (bennypowers) [reply elements/pf-select/pf-select.ts:289]: Pf5 removed that feature, and implementing it raises questions about tab order

File: elements/pf-select/docs/pf-select.md [hunk 5]
```diff
@@ -25,38 +25,36 @@ Focus on options using arrow keys or
...
+#### Checkbox input
+
+Multiple options can be selected. Any arrow keys work.
+<kbd>Shift</kbd> will toggling off multiple items.
+<kbd>Ctrl+A</kbd> will toggle selection on all items.
+
+{% htmlexample %}
+  {% renderFile "./elements/pf-select/demo/checkbox-input.html" %}
+{% endhtmlexample %}
+
```
[the remark this answers is not in the document]
Comment (bennypowers) [reply elements/pf-select/docs/pf-select.md:50]: some of these features are destined to be removed in v5, namely:

- Checkbox input with counts,
- Checkbox input no badge,
- Grouped single with filtering,
- Grouped checkbox input with filtering,
- Grouped checkbox input with filtering and placeholder text, and
- Grouped checkbox input with filtering and custom badging

So I'd rather skip them

File: elements/pf-select/docs/pf-select.md [hunk 6]
```diff
@@ -180,112 +99,24 @@ Multiple options, displayed as checkboxes, can be selected.
...
-    <pf-option value="Blue">Blue</pf-option>
-    <pf-option value="Green">Green</pf-option>
-    <pf-option value="Magenta">Magenta</pf-option>
-    <pf-option value="Orange">Orange</pf-option>
-    <pf-option value="Purple">Purple</pf-option>
-    <pf-option value="Pink">Pink</pf-option>
-    <pf-option value="Red">Red</pf-option>
-    <pf-option value="Yellow">Yellow</pf-option>
-  </pf-select>
+  {% renderFile "./elements/pf-select/demo/typeahead-multiple.html" %}
```
[the remark this answers is not in the document]
Comment (bennypowers) [reply elements/pf-select/docs/pf-select.md:107]: Let's leave multiple for when we implement typeahead, and skip features that will be removed in v5

**Comment by bennypowers** [comment]
https://deploy-preview-2677--patternfly-elements.netlify.app/components/select/
https://deploy-preview-2677--patternfly-elements.netlify.app/components/chip/

> * Chip group should be hidden if there are not chips, but it can be seen in DP

see f11741cc

**Review by nikkimk — CHANGES_REQUESTED** [review]
In [DP Single example](https://deploy-preview-2677--patternfly-elements.netlify.app/components/select/#single), if you select an item, both the disabled default item ("Select a title") and the newly selected item are selected. Only the newly selected item should be selected.

**Review by nikkimk — APPROVED** [review]
lawless genius twirls mustache

**Comment by bennypowers** [comment]
The failures in rh-tabs will be P.G. fixed in a subsequent PR

**Pull request merged**